query-algebrizer/tests/predicate.rs

```diff
@@ -0,0 +1,149 @@
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+extern crate mentat_core;
+extern crate mentat_query;
+extern crate mentat_query_algebrizer;
+extern crate mentat_query_parser;
+
+use std::collections::BTreeMap;
+
+use mentat_core::{
+    Attribute,
+    Entid,
+    Schema,
+    ValueType,
+    TypedValue,
+};
+
+use mentat_query_parser::{
+    parse_find_string,
+};
+
+use mentat_query::{
+    NamespacedKeyword,
+    PlainSymbol,
+    Variable,
+};
+
+use mentat_query_algebrizer::{
+    BindingError,
+    ConjoiningClauses,
+    ComputedTable,
+    EmptyBecause,
+    Error,
+    ErrorKind,
+    QueryInputs,
+    ValueTypeSet,
+    algebrize,
+    algebrize_with_inputs,
+};
+
+// These are helpers that tests use to build Schema instances.
+#[cfg(test)]
+fn associate_ident(schema: &mut Schema, i: NamespacedKeyword, e: Entid) {
+    schema.entid_map.insert(e, i.clone());
+    schema.ident_map.insert(i.clone(), e);
+}
+
+#[cfg(test)]
+fn add_attribute(schema: &mut Schema, e: Entid, a: Attribute) {
+    schema.schema_map.insert(e, a);
+}
+
+fn prepopulated_schema() -> Schema {
+    let mut schema = Schema::default();
+    associate_ident(&mut schema, NamespacedKeyword::new("foo", "date"), 65);
+    associate_ident(&mut schema, NamespacedKeyword::new("foo", "double"), 66);
+    add_attribute(&mut schema, 65, Attribute {
+        value_type: ValueType::Instant,
+        multival: false,
+        ..Default::default()
+    });
+    add_attribute(&mut schema, 66, Attribute {
+        value_type: ValueType::Double,
+        multival: false,
+        ..Default::default()
+    });
+    schema
+}
+
+fn bails(schema: &Schema, input: &str) -> Error {
+    let parsed = parse_find_string(input).expect("query input to have parsed");
+    algebrize(schema.into(), parsed).expect_err("algebrize to have failed")
+}
+
+fn bails_with_inputs(schema: &Schema, input: &str, inputs: QueryInputs) -> Error {
+    let parsed = parse_find_string(input).expect("query input to have parsed");
+    algebrize_with_inputs(schema, parsed, 0, inputs).expect_err("algebrize to have failed")
+}
+
+fn alg(schema: &Schema, input: &str) -> ConjoiningClauses {
+    let parsed = parse_find_string(input).expect("query input to have parsed");
+    algebrize(schema.into(), parsed).expect("algebrizing to have succeeded").cc
+}
+
+#[test]
+fn test_instant_predicates_require_instants() {
+    let schema = prepopulated_schema();
+
+    // You can't use a string for an inequality: this is a straight-up error.
+    let query = r#"[:find ?e
+                    :where
+                    [?e :foo/date ?t]
+                    [(> ?t "2017-06-16T00:56:41.257Z")]]"#;
+    match bails(&schema, query).0 {
+        ErrorKind::InvalidArgument(op, why, idx) => {
```

Comment:
https://github.com/brson/error-chain/issues/95
https://github.com/brson/error-chain/issues/134

We're also waiting for that so we can derive `Sync` and upgrade to the current version of `error_chain`!